### PR TITLE
`_WinSIGWINCH` implementation for non-Unix platforms

### PR DESCRIPTION
**Pull request by UltraStudioLTD**
Also **black** formatting and small optimisation

### CODE REVIEW

File: pytermgui/ansi_interface.py [hunk 1]
```diff
@@ -738,7 +741,7 @@ def translate_mouse(code: str, method: str) -> list[MouseEvent | None] | None:
             continue
 
         matches = list(pattern.finditer(sequence))
-        if matches == []:
+        if not matches:
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:744]: Good call on `matches == []` being a pretty odd choice, but I personally would prefer `if len(matches) == 0`, as it is much more explicit. Just saying `not matches` hides the type.

File: pytermgui/ansi_interface.py [hunk 2]
```diff
@@ -234,11 +224,24 @@ def _call_listener(self, event: int, data: Any) -> None:
             for callback in self._listeners[event]:
                 callback(data)
 
+    async def _WinSIGWINCH(self, check_again_in: float = 0.5) -> None:
+        """Asynchronous replacement for `signal`'s *SIGWINCH*"""
+        while True:
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:229]: The project always has a linebreak between the docstring and body of a function.

File: pytermgui/ansi_interface.py [hunk 3]
```diff
@@ -12,17 +12,25 @@
 import re
 import sys
 import signal
+import asyncio
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:15]: Import **asyncio** module since we need it to run `_WinSIGWINCH`

File: pytermgui/ansi_interface.py [hunk 4]
```diff
@@ -12,17 +12,25 @@
...
 
 from typing import Optional, Any, Union, Callable, Tuple
 from dataclasses import dataclass, fields
 from enum import Enum, auto as _auto
 from sys import stdout as _stdout
 from string import hexdigits
 from subprocess import run as _run, Popen as _Popen
-from os import name as _name, get_terminal_size, system
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:22]: Removed `get_terminal_size` function from **os** module since it would be better to import it from **shutil** which wraps it with helper functions with fallback capability

File: pytermgui/ansi_interface.py [hunk 5]
```diff
@@ -12,17 +12,25 @@
...
+from os import name as _name, system
+from shutil import get_terminal_size
+
+# This wraps `os.get_terminal_size` and adds few functions to
+#  try and get size if `os.get_terminal_size` fails
+#  and if all functions fail it fallbacks to (80, 24)
 
 from .input import getch
 
+_IS_NT = _name == "nt"
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:32]: Check if os is **Windows**

File: pytermgui/ansi_interface.py [hunk 6]
```diff
@@ -12,17 +12,25 @@
...
+_SYS_HAS_FRAME = hasattr(sys, "_getframe")
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:33]: Check if current **Python** implementation has frame get capability

File: pytermgui/ansi_interface.py [hunk 7]
```diff
@@ -12,17 +12,25 @@
...
 from sys import stdout as _stdout
 from string import hexdigits
 from subprocess import run as _run, Popen as _Popen
-from os import name as _name, get_terminal_size, system
+from os import name as _name, system
+from shutil import get_terminal_size
+
+# This wraps `os.get_terminal_size` and adds few functions to
+#  try and get size if `os.get_terminal_size` fails
+#  and if all functions fail it fallbacks to (80, 24)
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:28]: Good comment for the reviewer (moi), but we should remove this before merging. Nice replacement though!

File: pytermgui/ansi_interface.py [hunk 8]
```diff
@@ -191,25 +199,6 @@ def __call__(self, text: str, color: ColorType, reset_color: bool = True) -> str
...
-
-    try:
-        width, height = get_terminal_size()
-        return (width, height)
-
-    except OSError as error:
-        if error.errno in (9, 25):
-            return 0, 0
-
-        raise
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:210]: Worth pointing out that #6 had problems here. I would think the `shutil` implementation fixes those, however.

File: pytermgui/ansi_interface.py [hunk 9]
```diff
@@ -723,7 +726,7 @@ def translate_mouse(code: str, method: str) -> list[MouseEvent | None] | None:
             "34": MouseAction.RIGHT_CLICK,
             "35": MouseAction.RELEASE,
             "64": MouseAction.LEFT_DRAG,
-            "66": MouseAction.RIGHT_CLICK,
+            "66": MouseAction.RIGHT_DRAG,
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:729]: Fixed this typo

Comment (bczsalba) [inline pytermgui/ansi_interface.py:729]: Again, nice catch.

File: pytermgui/ansi_interface.py (hunk 1 above)
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:744]: Does same but more efficiently (and `pylint` doesn't nags me anymore)

File: pytermgui/ansi_interface.py [hunk 10]
```diff
@@ -12,17 +12,25 @@
...
 
 from typing import Optional, Any, Union, Callable, Tuple
 from dataclasses import dataclass, fields
 from enum import Enum, auto as _auto
 from sys import stdout as _stdout
 from string import hexdigits
 from subprocess import run as _run, Popen as _Popen
-from os import name as _name, get_terminal_size, system
+from os import name as _name, system
+from shutil import get_terminal_size
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:24]: Let's import `get_terminal_size` from **shutil**

File: pytermgui/ansi_interface.py [hunk 11]
```diff
@@ -191,25 +199,6 @@ def __call__(self, text: str, color: ColorType, reset_color: bool = True) -> str
...
-        width, height = get_terminal_size()
-        return (width, height)
-
-    except OSError as error:
-        if error.errno in (9, 25):
-            return 0, 0
-
-        raise
-
-
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:212]: We don't need this function since our imported *get_terminal_size* function from **shutil** does similar but better and more efficiently

File: pytermgui/ansi_interface.py [hunk 12]
```diff
@@ -223,9 +212,10 @@ def __init__(self) -> None:
         self.size: tuple[int, int] = self._get_size()
         self._listeners: dict[int, list[Callable[..., Any]]] = {}
 
-        # TODO: Windows doesn't support SIGWINCH, is there another alternative?
         if hasattr(signal, "SIGWINCH"):
             signal.signal(signal.SIGWINCH, self._update_size)
+        else:
+            asyncio.run(self._WinSIGWINCH())
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:218]: If **signal**'s module doesn't have *SIGWINCH* signal let's run our implementation/replacement

**Review by bczsalba — CHANGES_REQUESTED** [review]
Much better! Still some things to look at, but this is almost mergable.

File: pytermgui/ansi_interface.py [hunk 13]
```diff
@@ -234,11 +224,24 @@ def _call_listener(self, event: int, data: Any) -> None:
...
+        """Asynchronous replacement for `signal`'s *SIGWINCH*"""
+        while True:
+            n_width = get_terminal_size()
+            if n_width != self.size:
+                self._update_size(
+                    28,  # *SIGWINCH* is 28
+                    (sys._getframe(1) if _SYS_HAS_FRAME else None),
+                )
+            await asyncio.sleep(check_again_in)
+
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:237]: **_WinSIGWINCH** implementation

File: pytermgui/ansi_interface.py [hunk 14]
```diff
@@ -234,11 +224,24 @@ def _call_listener(self, event: int, data: Any) -> None:
...
     def _get_size(self) -> tuple[int, int]:
         """Get screen size while substracting the origin position"""
 
         # This always has len() == 2, but mypy can't see that.
-        return tuple(val - org for val, org in zip(screen_size(), self.origin))  # type: ignore
+        return tuple(
+            val - org for val, org in zip(get_terminal_size(), self.origin)
+        )  # type: ignore
```
Comment (UltraStudioLTD) [inline pytermgui/ansi_interface.py:244]: Little formatting for more readability

File: pytermgui/ansi_interface.py (hunk 1 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:744]: **pylint** warns me if I don't use `not matches` for some reason

Comment (bczsalba) [reply pytermgui/ansi_interface.py:744]: Which version of `pylint` and `black` are you running? There seems to be some odd behaviour, and `matches == []` is accepted on my machine.

File: pytermgui/ansi_interface.py (hunk 7 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:28]: I will remove it noe

**Comment by UltraStudioLTD** [comment]
I removed comments at import like you said

File: pytermgui/ansi_interface.py (hunk 8 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:210]: Believe it or not until I did this it wouldn't run on my Termux

File: pytermgui/ansi_interface.py (hunk 1 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:753]:
```pylint 2.12.2
astroid 2.9.3
Python 3.10.1 (main, Dec  9 2021, 15:28:13) [Clang 12.0.8 (https://android.googlesource.com/toolchain/llvm-project c935d99d7```

File: pytermgui/ansi_interface.py (hunk 2 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:236]: I formatted it to did that

File: pytermgui/ansi_interface.py (hunk 1 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:753]: It is latest version

File: pytermgui/ansi_interface.py [hunk 15]
```diff
@@ -75,7 +76,8 @@
 
 
 class Color:
-    """Class to store various color utilities
+    """
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:77]: I've gone to great detail in your previous PR about these changes not being necessary.

File: pytermgui/ansi_interface.py [hunk 16]
```diff
@@ -14,6 +14,7 @@
 import signal
 import asyncio
 
+from contextlib import contextmanager
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py]: This is outside of the scope of this PR.

File: pytermgui/ansi_interface.py [hunk 17]
```diff
@@ -221,6 +231,7 @@ def _call_listener(self, event: int, data: Any) -> None:
                 callback(data)
 
     async def _WinSIGWINCH(self, check_again_in: float = 0.5) -> None:
+
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:232]: What I meant with the docstring-body thing, is that you would have
```python
def function() -> None:
    """This is a docstring"""

    print("Hello")
```

File: pytermgui/ansi_interface.py [hunk 18]
```diff
@@ -282,24 +293,30 @@ def fill(self, color: int = 0, flush: bool = True) -> None:
...
+            termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, _)
+        try:
+            yield
+        finally:
+            if _IS_NT:
+                kernel32.SetConsoleMode(kernel32.GetStdHandle(-10), self._OldStdinMode)
+                kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), self._OldStdoutMode)
+            else:
+                termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, self._OldStdinMode)
+
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py]: Again, this is unrelated. Please move it to a different PR.

File: pytermgui/ansi_interface.py [hunk 19]
```diff
@@ -282,24 +293,30 @@ def fill(self, color: int = 0, flush: bool = True) -> None:
...
-    str_command = [str(c) for c in command]
-
-    if command[1] in waited_commands:
-        _run(str_command, check=True)
-        return
-
-    # A with statement here would result in `pass`, so is unnecessary.
-    _Popen(str_command)  # pylint: disable=consider-using-with
+# EDIT: removed `_tcup` function since we don't need it anymore.
+# I replaced it with ANSI sequences
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:299]: On some systems the ANSI sequences don't work. Those calls were needed.

File: pytermgui/ansi_interface.py [hunk 20]
```diff
@@ -316,29 +333,27 @@ def save_screen() -> None:
...
 
-    print("\x1b[?1049h")
+    _stdout.write("\x1b[?1049h")
 
 
 def unset_alt_buffer() -> None:
     """Return to main buffer from alt, restoring its original state"""
 
-    print("\x1b[?1049l")
+    _stdout.write("\x1b[?1049l")
```
Comment (bczsalba) [inline pytermgui/ansi_interface.py:336]: These have `print` as the function because the effects are supposed to be immediate, and you only write them to the screen one time per program execution in a normal case. Writing to `stdout` will make it buffered, which is not what these functions are meant to be.

Regardless, this is **outside of the scope**.

File: pytermgui/ansi_interface.py (hunk 1 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py]: I checked about `if len(matches) == 0` and it didn't yield any error about that so I will changed it to it

**Review by bczsalba — CHANGES_REQUESTED** [review]
Just stick to the windows `SIGWINCH` replacement, fix the things I mentioned and call it a day with this PR, please.

File: pytermgui/ansi_interface.py (hunk 15 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:77]: Just for readability

Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:77]: I remember it but when docstring is multiline it is better to make it like this. (Even other projects do like this for the sake of readability)

File: pytermgui/ansi_interface.py (hunk 20 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py]: I accidentally replaced commit code with code of planned commits

Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py]: I resolved it

File: pytermgui/ansi_interface.py (hunk 15 above)
Comment (bczsalba) [reply pytermgui/ansi_interface.py:77]: Read [the Google styleguide](https://google.github.io/styleguide/pyguide.html#382-modules). This project loosely follows this, in that we use the ending `"""` without a linebreak.

File: pytermgui/ansi_interface.py (hunk 19 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:299]: It's not system that defines ANSI support but terminal emulator itself. **tput** also uses escape sequences

File: pytermgui/ansi_interface.py (hunk 15 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py:77]: Alright

File: pytermgui/ansi_interface.py (hunk 17 above)
Comment (UltraStudioLTD) [reply pytermgui/ansi_interface.py]: I fixed it

**Comment by UltraStudioLTD** [comment]
Sorry for many commits. It looks like my editor and keyboard (I am coding on Android and sometimes it bugs) messed up formatting few times. Also when I was writing I had three projects open and I accidentally replaced to-commit code in few places but I fixed it. Please review it

**Review by bczsalba — COMMENTED** [review]
Better. Thank you. My previous comments still stand, from before all of the editor mishaps.

**Comment by UltraStudioLTD** [comment]
Also please note my comment on **tput** (even though I is outside of scopes of this PR I think it is still necessary for cross-platform functionality). Also I got Windows device and I will test this code on it too

**Comment by UltraStudioLTD** [comment]
> Better. Thank you. My previous comments still stand, from before all of the editor mishaps.

Do you mean that there are still things I should fix?

**Comment by bczsalba** [comment]
Yes. I have no idea where it all went to, but I can't find it anymore in this sea of comments. My biggest thing was that `_WinSIGWINCH` does not comply with out naming scheme. All function names should be verbs, and `snake_case`. Try something like `monitor_sigwinch` and mention that this is only used on Windows. I had more things to say, but I can't remember them nor find them anymore.

**Comment by UltraStudioLTD** [comment]
Ok. So last patch of this PR will be naming. In 1 minute

**Comment by UltraStudioLTD** [comment]
I renamed it to `_alt_sigwinch` since it is not only for Windows but for all platforms which don't support SIGWINCH signal and I think I am done with this PR.
Are there anything left?

**Comment by UltraStudioLTD** [comment]
Result should also fix

#6

File: pytermgui/ansi_interface.py (hunk 19 above)
Comment (bczsalba) [reply pytermgui/ansi_interface.py:292]: Look. I am aware of how these systems work; I managed to write this library somehow. My problem here is that I had a real compatibility issue with a device I was using to code on, where ANSI sequences did not work, but `tput` did. I am aware of the compatibility issues, but I still had that problem.

As I currently cannot test it (I do not have my old device at the moment) I'll let this pass.

**Comment by bczsalba** [comment]
> Result should also fix
> 
> #6

That was already fixed. The `tput` comments are still there in the file, but if I keep mentioning every single change I want you to make we are never getting anywhere. I'll merge this now, thank you for your work.

**Comment by UltraStudioLTD** [comment]
> The `tput` comments are still there in the file, but if I keep mentioning every single change I want you to make we are never getting anywhere. I'll merge this now, thank you for your work.

No problem but I removed that comments before commiting 🤔
Looks like Editor didn't save it. Forget it. I will fix it in next commit

**Comment by bczsalba** [comment]
So it turns out this broke installation on Windows. I removed it for the time being, and we may need to look towards the Windows APIs to get the functionality back.

**Comment by UltraStudioLTD** [comment]
As far as I know there is one way to get events using **Windows APIs**. It is done using [<kbd>ReadConsoleInput</kbd>](https://docs.microsoft.com/en-us/windows/console/readconsoleinput) to read **input buffer** and detect events in it. To detect resize event we need to capture [<kbd>WINDOW_BUFFER_SIZE_RECORD</kbd>](https://docs.microsoft.com/en-us/windows/console/window-buffer-size-record-str) and then compare it by previous capture.

**Comment by UltraStudioLTD** [comment]
I don't have currently access to **Windows** device but when I will have I will try to implement it.

**Comment by bczsalba** [comment]
> As far as I know there is one way to get events using **Windows APIs**. It is done using [ReadConsoleInput](https://docs.microsoft.com/en-us/windows/console/readconsoleinput) to read **input buffer** and detect events in it. To detect resize event we need to capture [WINDOW_BUFFER_SIZE_RECORD](https://docs.microsoft.com/en-us/windows/console/window-buffer-size-record-str) and then compare it by previous capture.

Yeah, that's the way I know it as well. I also don't really have a Windows dev machine at the moment, so I can't really develop it myself.

**Comment by UltraStudioLTD** [comment]
Which would be better? I personally think first is better

**Comment by bczsalba** [comment]
I think the first one is better, but as you said by involving Win APIs we are kind of forcing ourselves to redo a lot of other parts of the input system.

Doing a monitoring thread could work fine, but it caused issues when I tried it, same with the async loop. Probably just me doing the implementation badly, though.

**Comment by UltraStudioLTD** [comment]
> I think the first one is better, but as you said by involving Win APIs we are kind of forcing ourselves to redo a lot of other parts of the input system.
That's true, although I think it will simplify ``Windows``'s implementation since <kbd>ReadConsoleInput</kbd> also handles <kbd>MOUSE_EVENT_RECORD</kbd>s too. Also [**Textual**](https://github.com/Textualize/textual)'s [**win32**](https://github.com/Textualize/textual/blob/main/src/textual/drivers/win32.py) contains thread-based handler so why not implement it?

**Comment by bczsalba** [comment]
I don't want to depend on Textual. We probably will have to implement something similar at some point.

**Comment by UltraStudioLTD** [comment]
> I don't want to depend on Textual. We probably will have to implement something similar at some point.

I am not saying we should depend on **Textual**. I am saying that we can implement handler using idea of **Textual**'s handler. That doesn't mean depending on it.

**Comment by UltraStudioLTD** [comment]
Is changing/not changing architecture of project important? Just curious

**Comment by bczsalba** [comment]
It's not very important, the only reason I generally oppose it is it's a lot of work and with university I'm not sure I have the time for it. If others can pick the project up and do it I'm fully fine with it though, as I'm very uncertain about the library on non UNIX platforms.

**Comment by UltraStudioLTD** [comment]
> It's not very important, the only reason I generally oppose it is it's a lot of work and with the university, I'm not sure I have the time for it

I don't think that's a lot of work. It won't ever take 1 day to restructure the entire project and it's fine if you don't have time.

> If others can pick the project up and do it I'm fully fine with it though, as I'm very uncertain about the library on non UNIX platforms.

That's what contributors are for.
